Add $useCache argument to getModuleTranslation() to be able to work around getting wrong translations

marekjedrzejewski opened the pull request:
Prevents getting translation in wrong language.

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | When locale was specified when invoking getModuleTranslation one could get string from other language (not original string, or default lang) if there was no string in specified locale.
| Type?         | bug fix 
| Category?     | CO 
| BC breaks?    | maybe
| Deprecations? | no
| Fixed ticket? | Fixes #13024
| How to test?  | Inside issue #20495 there is simple module uploaded, which can be used for testing this change.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/20537)
<!-- Reviewable:end -->

marekjedrzejewski committed:
Use global $_MODULES only when $locale is not set

Prevents getting translation in wrong language.

matks commented on classes/Translate.php:
```diff
@@ -194,6 +194,8 @@ public static function getModuleTranslation(
         }
 
         if (empty($iso)) {
+            // using global $_MODULES only when $locale is not set to prevent getting translation in wrong language
```
Hi ! I am sorry I dont think this bugfix can be merged. It might solve your issue but it impacts the whole Translate class and I am quite sure there will be unwanted side-effects.

Can you explain the logic behind this PR ? We might find another way that solves your issue without introducing others.

Do you only set `$_MODULES` here because `getIsoByLocale` uses it ?

marekjedrzejewski replied:
As I've written in the Issue mentioned in this PR, when I use `$this->l()` with locale specified, and there's no string for this locale, I get string in previous language that had translation for this string.
Because `$_MODULES` variable stores it without any mention of language the string is in, I can't know if that's proper translation, or something carried over from previous function call, so only thing that seemed to fix it for me was not using global `$_MODULES` when locale was provided. There might be better way to do this, and the way I did it might introduce bugs elsewhere, but I made this PR anyway, to get the ball rolling.

marekjedrzejewski replied:
Other thing I thought about is adding another argument, for example `useCache`, that is true by default, but this seemed more like a workaround than real fix.

matks reviewed (changes requested):
See comment

marekjedrzejewski committed:
Add  to getModuleTranslation

marekjedrzejewski commented:
My previous attempt was very naive, and not working after I added more strings to the equation. I decided to go with extending getModuleTranslation with useCache argument and it works with my test module. Doesn't seem to break anything, but previous one also didn't when testing really small case.

Here's updated version of module that can be used for testing:
[langtest.zip](https://github.com/PrestaShop/PrestaShop/files/5068931/langtest.zip)

marekjedrzejewski committed:
Did commit -am by mistake. Revering

marekjedrzejewski reviewed (commented):
Should be 'Add $useCache to getModuleTranslation'. Bash betrayed me :(

marekjedrzejewski committed:
Add product page link and move download link

Order details page didn't have link to product page which, in my opinion, is something really helpful there.

marekjedrzejewski committed:
Merge branch 'PrestaShop:develop' into develop

hibatallahAouadni commented:
@marekjedrzejewski why did you close your PR?

marekjedrzejewski commented:
It might be automatically closed after merge to my fork, but it might be me not being sure it doesn't break anything, without anyone more versed in presta than me to look at it. Or both. I don't remember why exactly it is closed, but I ended up using workaround, which I wrote about [here](https://github.com/PrestaShop/PrestaShop/issues/13024#issuecomment-694131637).
